source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -18,17 +21,115 @@ MapdataCheckPoint::MapdataCheckPoint(const SData *data)
 void MapdataCheckPoint::read(EGG::Stream &stream) {
     m_left.read(stream);
     m_right.read(stream);
-    m_jugemIndex = stream.read_u8();
+    m_jugemIndex = stream.read_s8();
     m_checkArea = stream.read_s8();
-    m_prevPt = stream.read_u8();
-    m_nextPt = stream.read_u8();
+    m_prevPt = stream.read_s8();
+    m_nextPt = stream.read_s8();
+}
+
+/// @brief Calculates @ref m_nextPoints and @ref m_prevPoints from @ref m_nextPt and @ref m_prevPt.
+/// @details Also calculates the quadrilaterals for the next checkpoints, filling the fields of @ref
+/// LinkedCheckpoint for each.
+/// @addr{0x80515624}
+void MapdataCheckPoint::initCheckpointLinks(MapdataCheckPointAccessor &accessor, int id) {
+    m_id = id;
+    auto checkPathAccessor = CourseMap::Instance()->getCheckPathAccessor();
+
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group,
+    // it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_prevCount = 0;
+
+        for (auto [i, prevID] : std::views::enumerate(checkpath->prev())) {
+            if (prevID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *prev = checkPathAccessor->get(prevID);
+            m_prevPoints[i] = accessor.get(prev->end());
+            m_prevCount++;
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        m_prevCount++;
+    }
+
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_nextCount = 0;
+
+        for (auto [i, nextID] : std::views::enumerate(checkpath->next())) {
+            if (nextID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *next = checkPathAccessor->get(nextID);
+            m_nextPoints[i].checkpoint = accessor.get(next->start());
+            m_nextCount++;
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        m_nextCount++;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (size_t i = 0; i < 6; i++) {
+        if (i < m_nextCount) {
+            auto &nextLinked = m_nextPoints[i];
+            auto *nextPoint = nextLinked.checkpoint;
+
+            nextLinked.distance = (nextPoint->m_midpoint - m_midpoint).normalise();
+            nextLinked.p0diff =
+                    EGG::Vector2f(nextPoint->m_left.x - m_left.x, nextPoint->m_left.y - m_left.y);
```

Comment:
This is assuming that `Vector2f::operator-` exists. I don't remember if it does. If not, we should add it
```suggestion
            nextLinked.p0diff = nextPoint->m_left - m_left;
```